### ISSUE TITLE
csurf in SSR apps

### ISSUE THREAD

**Issue by jes490** [expressjs/csurf#185]
Hello!

I have some issues with csurf in SSR (nuxt.js in my case) app served as express middleware. 
To pass CSRF validation you need both _csrf cookie and hidden input CSRF token (or CSRF header in case of ajax). With each request csurf adds _csrf cookie to each response and all next requests will automatically add that cookie. All is good. 

But when you use SSR app, you can encounter following situation: 
- user clears all cookies (or cookie expires, it doesn't matter)
- user refreshes the page and makes request (lets name it request [**A**]
- csurf adds _csrf cookie to response of that request [**A**]
- after csurf request goes to the next middleware - SSR app (nuxt.js)
- nuxt.js in order to render the page requested in request [**A**] needs to make another ajax (axios)  request [**B**] even before request [**A**] is completed
- so now in request [**B**] there is no _csrf cookie because it was made from server before request [**A**] even completed and cookies are set
- as a result you will have 'Invalid csrf token' for that request [**B**]

If you refresh the page one more time - all will work, because _csrf cookie will be set. But every time you clear your cookies and refreshes for the first time situation above happens. 

---

How can I handle situations like this? I'm stuck basically, trying to solve this issue for 2 days now. Nothing comes to mind... 
Does someone have some suggestions/advises?

**Comment by rclement** [comment]
I stumble upon a similar issue using `nuxt` SSR + `csurf`, the csrf token cookie not being generated only on first load.

After a few days investigating, I found the culprit: `nuxt-i18n` was discarding all `Set-Cookie` headers when performing the SSR, only the first time when the redirection language cookie was not present in the request. Subsequent requests were embedding the redirection language cookie, thus not clearing the 'Set-Cookie' headers, the csrf token cookie being properly sent to the end of the response...

I submitted a PR upstream, it has been merged and released with `nuxt-i18n@^5.12.6`:
https://github.com/nuxt-community/nuxt-i18n/pull/327

If you are not using `nuxt-i18n`, then you might try using the `on-headers` package to generate the csrf token cookie, something like:

```
import onHeaders from 'on-headers'
...
app.use((req, res, next) => {
  onHeaders(res, function() {
    this.cookie('csrf-token', req.csrfToken(), {
      secure: production,
      httpOnly: false,
      sameSite: true
    })
  }
  next()
})
```
This is the technique used by the `cookie-session` package to generate the final session cookie(s) at the very end of the response.

Good luck!
